### PR TITLE
CONTRIBUTING: Add docs of the Architecture

### LINKED ISSUE #52: Document the internal flow of the software
Understanding how the internals of the software are organized, how logic is applied to data, and how code is grouped together would make it easier for interested people (like me) to make sense of the software and begin contributing fixes/features.

The `golangci-lint` project does a great job of this, and is what I have in mind. https://golangci-lint.run/contributing/architecture/

I saw a "how does it work?" section in the README, but it's a 30,000 foot view and I'm looking for a 1,000 foot view.

### PR DESCRIPTION
**Pull request by xescugc**
So it's can be easy to get into the project and know where everything is placed.

I think it has everything we may want no?

Closes #52

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
 find . -name '*-gray.svg' -type f -exec inkscape {} -w 72 -h 72 --export-filename resize/{}.png \; && find ./resize/ -type f -name '*.svg.png' -exec sh -c 'f="{}"; mv -- "$f" "${f%.svg.png}.png"' \;
 ```
 
+### Code Architecture
+
+In this section we'll explain the main packages if you want to contribute/read/understand how InfraMap does what it does.
```
Comment (tormath1) [inline CONTRIBUTING.md:110]:
```suggestion
In this section we'll explain the main packages if you want to contribute/read/understand how InfraMap works.
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+
+#### Providers
+
+The main idea behind InfraMap is to abstract each specific provider logic into a different implementation with a common interface
```
Comment (tormath1) [inline CONTRIBUTING.md:114]: This para is a bit redundant with himself. I would reword in in this way:

> The main idea behind InfraMap is to abstract each specific Terraform provider logic into a different implementation using a common interface: `provider.Provider`.
All this implementations are inside the `provider/` package.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+so then we can have a common logic that calls this interface without knowing which provider is using. All this interface implementations
+are inside the `provider/` package and being `provider.Provider` the main interface with all the needed methods for each provider.
+
+Not all the methods are needed to be implemented and that's why we have the `provider/nop_provider.go` which has an empty implementation
+of the `provider.Provider` interface.
+
+All the providers that we have implemented are defined on `provider/{provider-name}` for example `provider/aws/` which holds
```
Comment (tormath1) [inline CONTRIBUTING.md:121]: We could add a note regarding the `{provider-name}`: it needs to reflect the TF provider name.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+the implementation of the AWS provider.
+
+For more information on the provider interface we recommend to read the interface itself.
```
Comment (tormath1) [inline CONTRIBUTING.md:124]: We could add a doc link: https://pkg.go.dev/github.com/cycloidio/inframap@v0.3.0/provider?tab=doc#Provider

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+
+#### Generate and Prune
+
+The main entry point start on the `main.go` but it's basically delegated to the `cmd/`. The logic of each subcommand (`generate` and `prune`)
```
Comment (tormath1) [inline CONTRIBUTING.md:128]:
```suggestion
The main entry point starts with the `main.go` which actually calls the `cmd/`. The logic of each subcommand: `generate` and `prune`
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+is separated into different packages: `generate/` and `prune/` respectively.
```
Comment (tormath1) [inline CONTRIBUTING.md:129]:
```suggestion
is separated into eponymous packages.
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+
+**Prune**
+
+For the Prune what we do is use the `Provider.UsedAttributes` as a white list of the attributes that are actually needed and
```
Comment (tormath1) [inline CONTRIBUTING.md:133]:
```suggestion
For the Prune what we do is to use the `Provider.UsedAttributes` as a white list of the attributes that are actually needed,
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+drop the rest of the attributes and output the pruned version of the input.
```
Comment (tormath1) [inline CONTRIBUTING.md:134]:
```suggestion
drop the rest and output the pruned version of the input.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+
+**Generate**
+
+For the Generate it's a bit more complex. We want to have a common structure to operate with that it's independent of the input type (HCL or TFState),
```
Comment (tormath1) [inline CONTRIBUTING.md:138]:
```suggestion
For the Generate it's a bit more complex. We want to have a common structure to operate in order to be independent of the input type (HCL or TFState),
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+
+  tags = {
+    Name = "name"
+    type = "master"
+    role = "rds"
+  }
+}
+```
+
+This means that `aws_db_instance.application -> aws_db_instance.rds.id` will have an edege between them.
```
Comment (tormath1) [inline CONTRIBUTING.md:161]:
```suggestion
This means that `aws_db_instance.application -> aws_db_instance.rds.id` will have an edge between them.
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+
+This is not always enough so we have also have a `Provider.PreProcess` which from the JSON config (common JSON representation of the HCL or TFState configuration)
+can return specific edges to create between nodes.
```
Comment (tormath1) [inline CONTRIBUTING.md:164]: > an example can be found for the Google provider https://github.com/cycloidio/inframap/blob/074e1f9bc98fb99ede7d55404b16b4d9ff9cb1e1/provider/google/google.go#L167

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+
+Once we have the common Graph, then we start doing the operations which will be:
```
Comment (tormath1) [inline CONTRIBUTING.md:166]:
```suggestion
Once we have the common Graph, we start doing the operations which will be:
```

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+* Fix the direction of the edges, some of the first edges are wrong in terms of direction, so we fix them first. We do that by checking the config and asking the provider which are the right directions and fix them.
```
Comment (tormath1) [inline CONTRIBUTING.md:167]: > Using the `ResourceInOut` method, we fix the direction of the edges...

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+* Remove the nodes that are edges on the specific provider (ex: `aws_security_group` is an edge on AWS). We do that by checking which is the shortest connections between two nodes (provider nodes) on the Graph and then merge all the nodes in between. The merging works by merging each node one by one and reassigning all the edges to the new node with the right direction. To keep the right direction at the end of the merge, and successive merges, we keep a weigh that is used to know how many merges have been done and in which direction (read the code for more information).
+
+After all this we clean the end graph by removing edges that where not connected or merged, because they are just hanging on the Graph.
```
Comment (tormath1) [inline CONTRIBUTING.md:170]:
```suggestion
After all this we clean the end graph by removing not connected/merged edges because they are just hanging on the Graph.
```

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -110,3 +105,68 @@ You need to be in the directory in which the images are defined and have a `resi
...
+
+When all this operations are done then we Print the Graph.
```
Comment (tormath1) [inline CONTRIBUTING.md:172]:
```suggestion
When all this operations are done, we finally printe the Graph using the `printer/` package.
```

**Comment by xescugc** [comment]
Done

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -110,3 +105,66 @@ You need to be in the directory in which the images are defined and have a `resi
...
+  }
+}
+```
+
+This means that `aws_db_instance.application -> aws_db_instance.rds.id` will have an edge between them.
+
+This is not always enough so we have also have a `Provider.PreProcess` which from the JSON config (common JSON representation of the HCL or TFState configuration)
+can return specific edges to create between nodes. An example can be found for the Google provider https://github.com/cycloidio/inframap/blob/074e1f9bc98fb99ede7d55404b16b4d9ff9cb1e1/provider/google/google.go#L167
+
+Once we have the common Graph, we start doing the operations which will be:
```
Comment (tormath1) [inline CONTRIBUTING.md:164]:
```suggestion
Once we have the common Graph, we start doing the operations:
```

**Comment by xescugc** [comment]
Done

**Comment by tormath1** [comment]
OK for rebase !

**Comment by xescugc** [comment]
Done

**Review by tormath1 — APPROVED** [review]
